src/lazy/expanded/macro_evaluator.rs

```diff
@@ -606,17 +663,68 @@ impl<'data, D: LazyDecoder<'data>, M: MacroInvocation<'data, D>> MakeStringExpan
     }
 }
 
+// ===== Implementation of template macro expansion =====
+
+/// The evaluation state of a template expansion.
+#[derive(Clone, Debug)]
+pub struct TemplateExpansion<'top> {
+    // A reference to the template definition
+    template: TemplateMacroRef<'top>,
+    // The current 'step' of the expansion being processed.
+    step_index: usize,
+}
+
+impl<'top> TemplateExpansion<'top> {
+    pub fn new(template: TemplateMacroRef<'top>) -> Self {
+        Self {
+            template,
+            step_index: 0,
+        }
+    }
+
+    fn next<'data: 'top, D: LazyDecoder<'data>>(
+        &mut self,
+        context: EncodingContext<'top>,
+        environment: Environment<'top, 'data, D>,
+    ) -> IonResult<Option<ValueExpr<'top, 'data, D>>> {
+        let value_expr = match self.template.body().expressions().get(self.step_index) {
+            None => return Ok(None),
+            Some(expr) => expr,
+        };
+        self.step_index += 1;
+
+        let step = match value_expr {
+            TemplateBodyValueExpr::Element(e) => {
+                match e.value() {
+                    TemplateValue::List(range)
+                    | TemplateValue::SExp(range)
+                    | TemplateValue::Struct(range) => self.step_index += range.len(),
```

Comment:
🗺️ If the value in the template body is a container, it will be able to iterate over its child elements from the template body. This iterator can skip over them.